daemon/stop.go

```diff
@@ -38,52 +38,56 @@ func (daemon *Daemon) ContainerStop(name string, timeout *int) error {
-		// because if we can't stop the container by this point then
-		// it's probably because it's already stopped. Meaning, between
-		// the time of the IsRunning() call above and now it stopped.
-		// Also, since the err return will be environment specific we can't
-		// look for any particular (common) error that would indicate
-		// that the process is already dead vs something else going wrong.
-		// So, instead we'll give it up to 2 more seconds to complete and if
-		// by that time the container is still running, then the error
-		// we got is probably valid and so we force kill it.
-		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
-		defer cancel()
 
-		if status := <-container.Wait(ctx, containerpkg.WaitConditionNotRunning); status.Err() != nil {
-			logrus.Infof("Container failed to stop after sending signal %d to the process, force killing", stopSignal)
-			if err := daemon.killPossiblyDeadProcess(container, 9); err != nil {
-				return err
-			}
-		}
+	// 1. Send a stop signal
+	err := daemon.killPossiblyDeadProcess(container, stopSignal)
+	if err != nil {
+		wait = 2 * time.Second
 	}
 
-	// 2. Wait for the process to exit on its own
-	ctx := context.Background()
+	subCtx, cancel := context.WithCancel(ctx)
 	if seconds >= 0 {
-		var cancel context.CancelFunc
-		ctx, cancel = context.WithTimeout(ctx, time.Duration(seconds)*time.Second)
-		defer cancel()
+		// if seconds has been set, then set a timeout on the context so we dont wait forever
+		subCtx, cancel = context.WithTimeout(ctx, wait)
+	}
+	defer cancel()
+	if status := <-container.Wait(subCtx, containerpkg.WaitConditionNotRunning); status.Err() == nil {
+		// container did exit, so ignore any previous errors and return
+		return success()
 	}
 
-	if status := <-container.Wait(ctx, containerpkg.WaitConditionNotRunning); status.Err() != nil {
-		logrus.Infof("Container %v failed to exit within %d seconds of signal %d - using the force", container.ID, seconds, stopSignal)
-		// 3. If it doesn't, then send SIGKILL
-		if err := daemon.Kill(container); err != nil {
-			// Wait without a timeout, ignore result.
-			<-container.Wait(context.Background(), containerpkg.WaitConditionNotRunning)
-			logrus.Warn(err) // Don't return error because we only care that container is stopped, not what function stopped it
+	if err != nil {
+		// the container has still not exited, and the kill function errored, so log the error here:
+		logrus.WithError(err).WithField("container", container.ID).Errorf("Error sending stop (signal %d) to container", stopSignal)
+	}
+	logrus.WithField("container", container.ID).Infof("Container failed to exit within %d seconds of signal %d - using the force", seconds, stopSignal)
+
+	// Stop either failed or container didnt exit, so fallback to kill.
+	if err := daemon.Kill(container); err != nil {
+		// got a kill error, but give container 2 more seconds to exit just in case
+		subCtx, cancel := context.WithTimeout(ctx, 2*time.Second)
```

Comment:
I would bump this up to a higher value just because if the system is loaded this could have cause more issues.... maybe a minute?

I don't think we have the container lock here. Presuming I'm right waiting shouldn't cause any problems just so long as we eventually give up.

Reply:
At this point the kill signal has failed for some reason, waiting for 1 minute seems overly long to me. Reason being that a caller of this function may set a timeout for how long they are willing to wait at a higher level than this function. If they passed in 10s as the timeout (the default) to container stop it seems a bit excessive to wait for an entire minute after the kill signal has failed.

If you'd really like to wait longer here, perhaps something on the order of 10 seconds?

Reply:
@sparrc Currently the only way for the caller to set a timeout is from the passed in wait value. We could use that, but I think 2 seconds is too small and indefinite is too long.

Reply:
In ECS we implemented a timeout to avoid our code waiting forever on docker returning from the ContainerStop function call. In my opinion this function shouldn't wait much longer than what the user passes in via the `seconds` argument.

If we use `wait` again here, then this function will usually take around `wait` seconds, with the possibility to take `2 * wait` seconds. To me this is probably better than `wait + 60s`

I guess this is all a moot point anyways if you say that you have a change incoming to add the ability for the user to pass in a context anyways. I will amend this PR to use `wait` again here.

Reply:
That being said, in the above code we lower `wait` to 2s when SIGTERM fails, so this logic may start to get complicated unless you would like me to also increase the wait after SIGTERM failure to something else?